testing/sampletests/main.go

```diff
@@ -0,0 +1,245 @@
+	// Uses a set of regions instead of a slice to avoid duplication.
+	testRegionTags = map[string]map[string]struct{}{}
+
+	// Initialize the map values of testRegionTags.
+	for _, t := range testFileRanges {
+		for _, r := range t {
+			testRegionTags[r.testName] = map[string]struct{}{}
+		}
+	}
+
+	uniqueRegionTags := map[string]struct{}{}
+
+	// Iterate through every file.
+	// Note: implicitly, if a file isn't tested, neither are its regions. But,
+	// we walk through every file so we can get all region tags to compute
+	// region tag coverage.
+	err = filepath.Walk(".", func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+
+		if info.IsDir() {
+			return nil
+		}
+
+		path, err = filepath.Abs(path)
+		if err != nil {
+			return err
+		}
+
+		src, err := ioutil.ReadFile(path)
+		if err != nil {
+			return err
+		}
+
+		lines := strings.Split(string(src), "\n")
+		for lineNum, line := range lines {
+			if regions := startRe.FindAllStringSubmatch(line, -1); len(regions) > 0 {
+				region := regions[0][1]
+				uniqueRegionTags[region] = struct{}{}
+				for _, r := range testFileRanges[path] {
+					if r.start >= lineNum && r.end >= lineNum {
+						testRegionTags[r.testName][region] = struct{}{}
+					}
+				}
+			}
+			if regions := endRe.FindAllStringSubmatch(line, -1); len(regions) > 0 {
+				region := regions[0][1]
+				for _, r := range testFileRanges[path] {
+					if r.start >= lineNum && r.end >= lineNum {
+						testRegionTags[r.testName][region] = struct{}{}
+					}
+				}
+			}
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return uniqueRegionTags, testRegionTags, nil
+}
+
+// testCoverage returns a map from file path to a slice of testRanges, which
+// include the name of the test and the line numbers of the file path tested.
+//
+// testCoverage only looks at direct function calls from the given test, it
+// does not look at transitive calls. This may lead to missing some region tags.
+func testCoverage() (map[string][]testRange, error) {
+	result := map[string][]testRange{}
+
+	config := &packages.Config{
+		Mode:  packages.NeedName | packages.NeedSyntax | packages.NeedTypes | packages.NeedTypesInfo,
+		Tests: true,
+	}
+
+	pkgs, err := packages.Load(config, "./...")
+	if err != nil {
+		return nil, fmt.Errorf("packages.Load: %v", err)
+	}
+	for _, pkg := range pkgs {
+		for id, obj := range pkg.TypesInfo.Defs {
+			if !strings.HasPrefix(id.Name, "Test") {
+				continue
+			}
+			var file *ast.File
+			for _, f := range pkg.Syntax {
+				if f.Pos() <= id.Pos() && id.Pos() <= f.End() {
+					file = f
+				}
+			}
+			if file == nil {
+				return nil, fmt.Errorf("file for %q not found", id.Name)
+			}
+			path, exact := astutil.PathEnclosingInterval(file, id.Pos(), id.End())
+			if !exact {
+				return nil, fmt.Errorf("PathEnclosingInterval got not exact path for %q in %v", id.Name, file)
+			}
+			ast.Inspect(path[1], func(node ast.Node) bool {
```

Comment:
I'm assuming this `ast` utility won't translate to other languages?

Reply:
No, it's specific to Go. Other languages might have other AST parsers, but this one won't work.